crates/bevy_pbr/src/light.rs

```diff
@@ -568,19 +784,135 @@ pub(crate) fn assign_lights_to_clusters(
+                    is_orthographic,
+                    inverse_view_transform,
+                    camera.projection_matrix,
+                    &light_sphere,
+                );
+
+                // since we won't adjust z slices we can calculate exact number of slices required in z dimension
+                let z_cluster_min = view_z_to_z_slice(
+                    cluster_factors,
+                    cluster_dimensions.z as f32,
+                    light_aabb_min.z,
+                    is_orthographic,
+                );
+                let z_cluster_max = view_z_to_z_slice(
+                    cluster_factors,
+                    cluster_dimensions.z as f32,
+                    light_aabb_max.z,
+                    is_orthographic,
+                );
+                let z_count =
+                    z_cluster_min.max(z_cluster_max) - z_cluster_min.min(z_cluster_max) + 1;
+
+                // calculate x/y count using floats to avoid overestimating counts due to large initial tile sizes
+                let xy_min = light_aabb_min.xy();
+                let xy_max = light_aabb_max.xy();
+                // multiply by 0.5 to move from [-1,1] to [-0.5, 0.5], max extent of 1 in each dimension
+                let xy_count = (xy_max - xy_min)
+                    * 0.5
+                    * Vec2::new(cluster_dimensions.x as f32, cluster_dimensions.y as f32);
+
+                // add up to 2 to each axis to account for overlap
+                let x_overlap = if xy_min.x <= -1.0 { 0.0 } else { 1.0 }
+                    + if xy_max.x >= 1.0 { 0.0 } else { 1.0 };
+                let y_overlap = if xy_min.y <= -1.0 { 0.0 } else { 1.0 }
+                    + if xy_max.y >= 1.0 { 0.0 } else { 1.0 };
+                cluster_index_estimate +=
+                    (xy_count.x + x_overlap) * (xy_count.y + y_overlap) * z_count as f32;
+            }
+
+            if cluster_index_estimate > max_indices as f32 {
+                // scale x and y cluster count to be able to fit all our indices
+
+                // we take the ratio of the actual indices over the index estimate.
+                // this not not guaranteed to be small enough due to overlapped tiles, but
+                // the conservative estimate is more than sufficient to cover the
+                // difference
+                let index_ratio = max_indices as f32 / cluster_index_estimate as f32;
+                let xy_ratio = index_ratio.sqrt();
+
+                cluster_dimensions.x =
+                    ((cluster_dimensions.x as f32 * xy_ratio).floor() as u32).max(1);
+                cluster_dimensions.y =
+                    ((cluster_dimensions.y as f32 * xy_ratio).floor() as u32).max(1);
+            }
+        }
+
+        update_clusters(
+            screen_size_u32,
+            camera,
+            cluster_dimensions,
+            &mut clusters,
+            first_slice_depth,
+            far_z,
+        );
```

Comment:
```suggestion
        update_clusters(
            screen_size_u32,
            camera,
            cluster_dimensions,
            &mut clusters,
            first_slice_depth,
            far_z,
        );
        // NOTE: This is here to avoid bugs in future due to update_clusters() having updated clusters.axis_slices
        // but cluster_dimensions has a different configuration.
        cluster_dimensions = clusters.axis_slices;
```